In [2]:
import sys
sys.path.append('..')

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datetime import timedelta, datetime, date
import os
from utils import data_paths, load_config
from pathlib import Path
from nltk.metrics import edit_distance #(Levenshtein)
import pycountry
import math

# Estimating The Infected Population From Deaths
> Estimating the number of infected people by country based on the number of deaths and case fatality rate. 

- comments: true
- author: Joao B. Duarte
- categories: [growth, compare, interactive, estimation]
- hide: false
- image: images/covid-estimate-infections.png
- permalink: /covid-infected/
- toc: true

In [4]:
LOCAL_FILES=True
#jupyter or script
IS_SCRIPT = False

In [5]:
os.getcwd()

'/mnt/963GB/Data/Python/ACode/medical/covid19/forecaster/covidforecaster/tools'

In [6]:
if IS_SCRIPT:
    RUN_PATH = Path(os.path.realpath(__file__))
    DATA_PARENT = RUN_PATH.parent.parent
else:
    #for jupyter
    cw=!pwd
    RUN_PATH = Path(cw[0])
    DATA_PARENT = RUN_PATH.parent

In [7]:
if IS_SCRIPT:
    csse_data = data_paths('tools/csse_data_paths.yml')
else:
    csse_data = data_paths('csse_data_paths.yml')

In [8]:
if LOCAL_FILES:
    confirmed_url=csse_data.get("csse_ts_local", {}).get('confirmed', {})
    deaths_url=csse_data.get("csse_ts_local", {}).get('deaths', {})
    recovered_url=csse_data.get("csse_ts_local", {}).get('recovered', {})
    
    confirmed_url = str(DATA_PARENT/confirmed_url)
    deaths_url = str(DATA_PARENT/deaths_url)
    recovered_url = str(DATA_PARENT/recovered_url)
else:
    confirmed_url=csse_data.get("csse_ts_global", {}).get('confirmed', {})
    deaths_url=csse_data.get("csse_ts_global", {}).get('deaths', {})
    recovered_url=csse_data.get("csse_ts_global", {}).get('recovered', {})

In [9]:
### UN stats

In [10]:
df_un_pop_density_info=pd.read_csv(DATA_PARENT/'data/un/df_un_pop_density_info.csv')
df_un_urban_growth_info=pd.read_csv(DATA_PARENT/'data/un/urban_growth_info.csv')
df_un_health_info=pd.read_csv(DATA_PARENT/'data/un/df_un_health_info.csv')
df_un_tourism_info=pd.read_csv(DATA_PARENT/'data/un/df_un_tourism_info.csv')
df_un_gdp_info=pd.read_csv(DATA_PARENT/'data/un/df_un_gdp_info.csv')
df_un_edu_info=pd.read_csv(DATA_PARENT/'data/un/df_un_edu_info.csv')
df_un_pop_growth_info=pd.read_csv(DATA_PARENT/'data/un/df_un_pop_growth_info.csv')
df_un_gdrp_rnd_info=pd.read_csv(DATA_PARENT/'data/un/df_un_gdrp_rnd_info.csv')
df_un_education_info=pd.read_csv(DATA_PARENT/'data/un/df_un_education_info.csv')
df_un_sanitation_info=pd.read_csv(DATA_PARENT/'data/un/df_un_sanitation_info.csv')

df_un_health_expenditure_info=pd.read_csv(DATA_PARENT/'data/un/df_un_health_expenditure_info.csv')
df_un_immigration_info=pd.read_csv(DATA_PARENT/'data/un/df_un_immigration_info.csv')
df_un_trading_info=pd.read_csv(DATA_PARENT/'data/un/df_un_trading_info.csv')
df_un_land_info=pd.read_csv(DATA_PARENT/'data/un/df_un_land_info.csv')

In [11]:
df_un_health_info.head()
#Health personnel: Pharmacists (per 1000 population)

Region/Country/Area      Country  Year  \
0                    4  Afghanistan  2001   
1                    4  Afghanistan  2001   
2                    4  Afghanistan  2001   
3                    4  Afghanistan  2001   
4                    4  Afghanistan  2005   

                                              Series   Value  Footnotes  \
0              Health personnel: Physicians (number)  4104.0        NaN   
1  Health personnel: Physicians (per 1000 populat...     0.2        NaN   
2             Health personnel: Pharmacists (number)   525.0        NaN   
3  Health personnel: Pharmacists (per 1000 popula...     0.0        NaN   
4             Health personnel: Pharmacists (number)   900.0        NaN   

                                              Source  
0  World Health Organisation (WHO), Geneva, WHO G...  
1  World Health Organisation (WHO), Geneva, WHO G...  
2  World Health Organisation (WHO), Geneva, WHO G...  
3  World Health Organisation (WHO), Geneva, WHO G...  
4  World Health Organisation (WHO), Geneva, WHO G...

In [12]:
df_un_trading_info.tail(n=20)
#column Major trading partner 1 (% of exports)
#Major trading partner 1 (% of exports)
#Major trading partner 2 (% of exports)
#Major trading partner 3 (% of exports)

Region/Country/Area   Country  Year  \
2834                  894    Zambia  2018   
2835                  894    Zambia  2005   
2836                  894    Zambia  2010   
2837                  894    Zambia  2018   
2838                  716  Zimbabwe  2005   
2839                  716  Zimbabwe  2010   
2840                  716  Zimbabwe  2018   
2841                  716  Zimbabwe  2005   
2842                  716  Zimbabwe  2010   
2843                  716  Zimbabwe  2018   
2844                  716  Zimbabwe  2005   
2845                  716  Zimbabwe  2010   
2846                  716  Zimbabwe  2018   
2847                  716  Zimbabwe  2005   
2848                  716  Zimbabwe  2010   
2849                  716  Zimbabwe  2018   
2850                  716  Zimbabwe  2018   
2851                  716  Zimbabwe  2005   
2852                  716  Zimbabwe  2010   
2853                  716  Zimbabwe  2018   

                                      Series  \
2834  Major trading partner 3 (% of exports)   
2835  Major trading partner 3 (% of imports)   
2836  Major trading partner 3 (% of imports)   
2837  Major trading partner 3 (% of imports)   
2838  Major trading partner 1 (% of exports)   
2839  Major trading partner 1 (% of exports)   
2840  Major trading partner 1 (% of exports)   
2841  Major trading partner 1 (% of imports)   
2842  Major trading partner 1 (% of imports)   
2843  Major trading partner 1 (% of imports)   
2844  Major trading partner 2 (% of exports)   
2845  Major trading partner 2 (% of exports)   
2846  Major trading partner 2 (% of exports)   
2847  Major trading partner 2 (% of imports)   
2848  Major trading partner 2 (% of imports)   
2849  Major trading partner 2 (% of imports)   
2850  Major trading partner 3 (% of exports)   
2851  Major trading partner 3 (% of imports)   
2852  Major trading partner 3 (% of imports)   
2853  Major trading partner 3 (% of imports)   

     Major trading partner 1 (% of exports)  \
2834                  Dem.Rep. of the Congo   
2835                                  China   
2836                                  China   
2837                                  China   
2838                           South Africa   
2839                           South Africa   
2840                           South Africa   
2841                           South Africa   
2842                           South Africa   
2843                           South Africa   
2844                   United Arab Emirates   
2845                   United Arab Emirates   
2846                   United Arab Emirates   
2847                              Singapore   
2848                              Singapore   
2849                              Singapore   
2850                              Areas nes   
2851                                  China   
2852                                  China   
2853                                  China   

     Major trading partner 1 (% of exports) footnote    Value Footnotes  \
2834                                             NaN   9.5418       NaN   
2835                                             NaN   3.3408       NaN   
2836                                             NaN   5.4446       NaN   
2837                                             NaN  13.6411       NaN   
2838                                             NaN  41.4877       NaN   
2839                                             NaN  54.2168       NaN   
2840                                             NaN  51.4752       NaN   
2841                                             NaN  14.9645       NaN   
2842                                             NaN  48.0266       NaN   
2843                                             NaN  39.2800       NaN   
2844                                             NaN   0.7012       NaN   
2845                                             NaN  10.2893       NaN   
2846                                             NaN  18.1383       NaN   
2847         

In [13]:
df_population_density=df_un_pop_density_info.loc[df_un_pop_density_info['Series'] == 'Population density']

In [14]:
df_population_density.tail(n=50)
#Population aged 60+ years old (percentage)
#Population density
#Population mid-year estimates (millions)

Region/Country/Area               Country  Year              Series  \
4417                  788               Tunisia  2017  Population density   
4425                  788               Tunisia  2019  Population density   
4432                  792                Turkey  2005  Population density   
4439                  792                Turkey  2010  Population density   
4446                  792                Turkey  2017  Population density   
4454                  792                Turkey  2019  Population density   
4461                  800                Uganda  2005  Population density   
4468                  800                Uganda  2010  Population density   
4475                  800                Uganda  2017  Population density   
4483                  800                Uganda  2019  Population density   
4490                  804               Ukraine  2005  Population density   
4497                  804               Ukraine  2010  Population density   
4504                  804               Ukraine  2017  Population density   
4512                  804               Ukraine  2019  Population density   
4519                  784  United Arab Emirates  2005  Population density   
4526                  784  United Arab Emirates  2010  Population density   
4533                  784  United Arab Emirates  2017  Population density   
4541                  784  United Arab Emirates  2019  Population density   
4548                  826        United Kingdom  2005  Population density   
4555                  826        United Kingdom  2010  Population density   
4562                  826        United Kingdom  2017  Population density   
4570                  826        United Kingdom  2019  Population density   
4577                  834              Tanzania  2005  Population density   
4584                  834              Tanzania  2010  Population density   
4591                  834              Tanzania  2017  Population density   
4599                  834              Tanzania  2019  Population density   
4606                  858               Uruguay  2005  Population density   
4613                  858               Uruguay  2010  Population density   
4620                  858               Uruguay  2017  Population density   
4628                  858               Uruguay  2019  Population density   
4635                  860            Uzbekistan  2005  Population density   
4642                  860            Uzbekistan  2010  Population density   
4649                  860            Uzbekistan  2017  Population density   
4657                  860            Uzbekistan  2019  Population density   
4664                  862             Venezuela  2005  Population density   
4671                  862             Venezuela  2010  Population density   
4678                  862             Venezuela  2017  Population density   
4686                  862             Venezuela  2019  Population density   
4693                  704               Vietnam  2005  Population density   
4700                  704               Vietnam  2010  Population density   
4707                  704               Vietnam  2017  Population density   
4715                  704               Vietnam  2019  Population density   
4722                  894                Zambia  2005  Population density   
4729                  894                Zambia  2010  Population density   
4736                  894                Zambia  2017  Population density   
4744                  894                Zambia  2019  Population density   
4751                  716              Zimbabwe  2005  Population density   
4758                  716              Zimbabwe  2010  Population density   
4765                  716              Zimbabwe  2017  Population density   
4773                  716              Zimbabwe  2019  Population density   

         Value            Footnotes  \
4417   73.5932                  NaN   


In [15]:

df_population_density.loc[df_population_density.groupby('Country')['Year'].idxmax()]

Region/Country/Area             Country  Year              Series  \
28                      4         Afghanistan  2019  Population density   
57                      8             Albania  2019  Population density   
86                     12             Algeria  2019  Population density   
106                    20             Andorra  2019  Population density   
135                    24              Angola  2019  Population density   
...                   ...                 ...   ...                 ...   
4686                  862           Venezuela  2019  Population density   
4715                  704             Vietnam  2019  Population density   
4143                  275  West Bank and Gaza  2019  Population density   
4744                  894              Zambia  2019  Population density   
4773                  716            Zimbabwe  2019  Population density   

         Value                  Footnotes  \
28     58.2694                        NaN   
57    105.1430                        NaN   
86     18.0763                        NaN   
106   164.1319                        NaN   
135    25.5276                        NaN   
...        ...                        ...   
4686   32.3290                        NaN   
4715  311.0978                        NaN   
4143  827.4784  Including East Jerusalem.   
4744   24.0265                        NaN   
4773   37.8583                        NaN   

                                                 Source Unnamed: 7  
28    United Nations Population Division, New York, ...        NaN  
57    United Nations Population Division, New York, ...        NaN  
86    United Nations Population Division, New York, ...        NaN  
106   United Nations Population Division, New York, ...        NaN  
135   United Nations Population Division, New York, ...        NaN  
...                                                 ...        ...  
4686  United Nations Population Division, New York, ...        NaN  
4715  United Nations Population Division, New York, ...        NaN  
4143  United Nations Population Division, New York, ...        NaN  
4744  United Nations Population Division, New York, ...        NaN  
4773  United Nations Population Division, New York, ...        NaN  

[166 rows x 8 columns]

In [16]:
df_population_density

Region/Country/Area      Country  Year              Series     Value  \
6                       4  Afghanistan  2005  Population density   39.2952   
13                      4  Afghanistan  2010  Population density   44.7041   
20                      4  Afghanistan  2017  Population density   55.5956   
28                      4  Afghanistan  2019  Population density   58.2694   
35                      8      Albania  2005  Population density  112.6573   
...                   ...          ...   ...                 ...       ...   
4744                  894       Zambia  2019  Population density   24.0265   
4751                  716     Zimbabwe  2005  Population density   31.2180   
4758                  716     Zimbabwe  2010  Population density   32.8234   
4765                  716     Zimbabwe  2017  Population density   36.8013   
4773                  716     Zimbabwe  2019  Population density   37.8583   

     Footnotes                                             Source Unnamed: 7  
6          NaN  United Nations Population Division, New York, ...        NaN  
13         NaN  United Nations Population Division, New York, ...        NaN  
20         NaN  United Nations Population Division, New York, ...        NaN  
28         NaN  United Nations Population Division, New York, ...        NaN  
35         NaN  United Nations Population Division, New York, ...        NaN  
...        ...                                                ...        ...  
4744       NaN  United Nations Population Division, New York, ...        NaN  
4751       NaN  United Nations Population Division, New York, ...        NaN  
4758       NaN  United Nations Population Division, New York, ...        NaN  
4765       NaN  United Nations Population Division, New York, ...        NaN  
4773       NaN  United Nations Population Division, New York, ...        NaN  

[680 rows x 8 columns]

In [17]:
### Freedom House stats

In [18]:
#Freedon House stats
def country_freedom():
    global_freedom = str(DATA_PARENT/'data/freedom_house/Global_Freedom.csv')
    df_global_free = pd.read_csv(global_freedom)
    internet_freedom = str(DATA_PARENT/'data/freedom_house/Internet_Freedom.csv')
    df_internet_free = pd.read_csv(internet_freedom)
    return df_global_free, df_internet_free
df_global_freedom, df_internet_freedom = country_freedom()

In [19]:
#csse countries
df_deaths = pd.read_csv(deaths_url, error_bad_lines=False)
df_confirmed = pd.read_csv(confirmed_url, error_bad_lines=False)
df_recovered = pd.read_csv(recovered_url, error_bad_lines=False)
csse_countries = []
for df in [df_deaths, df_confirmed, df_recovered]:
    c = set(df["Country/Region"].unique())
    csse_countries.append(c)
csse_countries = [item for sublist in csse_countries for item in sublist]
csse_countries = list(set(csse_countries))

## CSSE

In [20]:
# Get data on deaths D_t
df_deaths = pd.read_csv(deaths_url, error_bad_lines=False)
df_deaths = df_deaths.drop(columns=["Lat", "Long"])
df_deaths = df_deaths.melt(id_vars= ["Province/State", "Country/Region"])
df_deaths = pd.DataFrame(df_deaths.groupby(['Country/Region', "variable"]).sum())
df_deaths.reset_index(inplace=True)  
df_deaths = df_deaths.rename(columns={"Country/Region": "location", "variable": "date", "value": "total_deaths"})
df_deaths['date'] =pd.to_datetime(df_deaths.date)
df_deaths = df_deaths.sort_values(by = "date")
df_deaths.loc[df_deaths.location == "US","location"] = "United States"
df_deaths.loc[df_deaths.location == "Korea, South","location"] = "South Korea"

In [21]:
#confirmed

In [22]:
df_confirmed = pd.read_csv(confirmed_url, error_bad_lines=False)
df_confirmed = df_confirmed.drop(columns=["Lat", "Long"])
df_confirmed = df_confirmed.melt(id_vars= ["Province/State", "Country/Region"])
df_confirmed = pd.DataFrame(df_confirmed.groupby(['Country/Region', "variable"]).sum())
df_confirmed.reset_index(inplace=True)  
df_confirmed = df_confirmed.rename(columns={"Country/Region": "location", "variable": "date", "value": "total_cases"})
df_confirmed['date'] =pd.to_datetime(df_confirmed.date)
df_confirmed = df_confirmed.sort_values(by = "date")
df_confirmed.loc[df_confirmed.location == "US","location"] = "United States"
df_confirmed.loc[df_confirmed.location == "Korea, South","location"] = "South Korea"

In [23]:
df_confirmed.head()

location       date  total_cases
0                 Afghanistan 2020-01-22            0
3685                 Eswatini 2020-01-22            0
6432                    Libya 2020-01-22            0
1407   Bosnia and Herzegovina 2020-01-22            0
11256    United Arab Emirates 2020-01-22            0

In [24]:
df_final = pd.merge(df_deaths,
                 df_confirmed)

In [25]:
df_final.head()

location       date  total_deaths  total_cases
0             Afghanistan 2020-01-22             0            0
1                Eswatini 2020-01-22             0            0
2                   Libya 2020-01-22             0            0
3  Bosnia and Herzegovina 2020-01-22             0            0
4    United Arab Emirates 2020-01-22             0            0

In [26]:
df_final["CFR"] = df_final["total_deaths"]/df_final["total_cases"]
df_final["total_infected"] = np.NaN
df_final = df_final.sort_values(by = ['location', 'date'])
df_final = df_final.reset_index(drop = True)

In [27]:
df_un_pop_per_country=pd.read_csv(DATA_PARENT/'data/un/df_un_pop_per_country_info.csv')

In [28]:
def get_country_list(pop_cutoff=5.0):
    pop_nmill=df_un_pop_per_country.loc[df_un_pop_per_country['Value'] >= pop_cutoff]
    countries_n_plus=pop_nmill.Country.tolist()
    return countries_n_plus

In [29]:
csse_countries.sort()
csse_countries

['Afghanistan',
 'Albania',
 'Algeria',
 'Andorra',
 'Angola',
 'Antigua and Barbuda',
 'Argentina',
 'Armenia',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Bahamas',
 'Bahrain',
 'Bangladesh',
 'Barbados',
 'Belarus',
 'Belgium',
 'Belize',
 'Benin',
 'Bhutan',
 'Bolivia',
 'Bosnia and Herzegovina',
 'Brazil',
 'Brunei',
 'Bulgaria',
 'Burkina Faso',
 'Burma',
 'Cabo Verde',
 'Cambodia',
 'Cameroon',
 'Canada',
 'Central African Republic',
 'Chad',
 'Chile',
 'China',
 'Colombia',
 'Congo (Brazzaville)',
 'Congo (Kinshasa)',
 'Costa Rica',
 "Cote d'Ivoire",
 'Croatia',
 'Cuba',
 'Cyprus',
 'Czechia',
 'Denmark',
 'Diamond Princess',
 'Djibouti',
 'Dominica',
 'Dominican Republic',
 'Ecuador',
 'Egypt',
 'El Salvador',
 'Equatorial Guinea',
 'Eritrea',
 'Estonia',
 'Eswatini',
 'Ethiopia',
 'Fiji',
 'Finland',
 'France',
 'Gabon',
 'Gambia',
 'Georgia',
 'Germany',
 'Ghana',
 'Greece',
 'Grenada',
 'Guatemala',
 'Guinea',
 'Guinea-Bissau',
 'Guyana',
 'Haiti',
 'Holy See',
 'Honduras',


In [30]:
csse_countries=list(map(lambda x: x if x != 'Korea, South' else "South Kores", csse_countries))

In [31]:
countries_n_plus = get_country_list(pop_cutoff=5.0)

In [32]:
for j in countries_n_plus:
    if not j in csse_countries:
        print(j)

South Korea


In [33]:
for j in countries_n_plus:
    for i in df_final["date"].unique()[0:-8]:
        numer =  df_final.loc[(df_final.date == i + np.timedelta64(8, 'D')) & (df_final.location == j), "total_deaths"].iloc[0]
        denom = df_final.loc[(df_final.date == i + np.timedelta64(8, 'D')) & (df_final.location == j), "CFR"].iloc[0]
        df_final.loc[(df_final.date == i) & (df_final.location == j), "total_infected"] = numer/denom

/home/walle/.virtualenvs/covid_dashboard/lib/python3.6/site-packages/ipykernel_launcher.py:5: RuntimeWarning: invalid value encountered in true_divide
  """


In [34]:
df_final.head()

location       date  total_deaths  total_cases  CFR  total_infected
0  Afghanistan 2020-01-22             0            0  NaN             NaN
1  Afghanistan 2020-01-23             0            0  NaN             NaN
2  Afghanistan 2020-01-24             0            0  NaN             NaN
3  Afghanistan 2020-01-25             0            0  NaN             NaN
4  Afghanistan 2020-01-26             0            0  NaN             NaN

In [35]:
# Estimate growth rate of infected, g        
df_final['infected_g'] = np.log(df_final['total_infected'])
df_final['infected_g'] = df_final['infected_g'].diff() 

In [36]:
# Estimate number of infected given g
today = df_final.date.iloc[-1]
for j in countries_n_plus:
    for i in range(7,-1,-1):
        df_final.loc[(df_final.location == j) & (df_final.date == today - timedelta(i)), "total_infected"] = df_final.loc[df_final.location == j, "total_infected"].iloc[-i-2]*(1+df_final.loc[df_final.location == j, "infected_g"].aggregate(func = "mean"))

In [37]:
data_pc = df_final[['location', 'date', 'total_infected']].copy()

In [38]:
data_countries = []
data_countries_pc = []

In [39]:
for i in countries_n_plus:
    data_pc.loc[data_pc.location == i,"total_infected"] = data_pc.loc[data_pc.location == i,"total_infected"]

In [40]:
# Get each country time series
filter1 = data_pc["total_infected"] > 1
for i in countries_n_plus:
    filter_country = data_pc["location"]== i
    data_countries_pc.append(data_pc[filter_country & filter1])      

In [41]:
len(data_countries_pc)

105

In [42]:
data_countries_pc[0]

location       date  total_infected
52  Afghanistan 2020-03-14       40.000000
53  Afghanistan 2020-03-15       40.000000
54  Afghanistan 2020-03-16       74.000000
55  Afghanistan 2020-03-17       84.000000
56  Afghanistan 2020-03-18       94.000000
57  Afghanistan 2020-03-19      110.000000
58  Afghanistan 2020-03-20      110.000000
59  Afghanistan 2020-03-21      128.546017
60  Afghanistan 2020-03-22      150.218895
61  Afghanistan 2020-03-23      175.545823
62  Afghanistan 2020-03-24      205.142876
63  Afghanistan 2020-03-25      239.729996
64  Afghanistan 2020-03-26      280.148509
65  Afghanistan 2020-03-27      327.381590
66  Afghanistan 2020-03-28      382.578176

## Estimated Infected Population By Country

by days since outbreak

In [43]:
# Lastest Country Estimates  
label = 'Total_Infected'
temp = pd.concat([x.copy() for x in data_countries_pc]).loc[lambda x: x.date >= '3/1/2020']

In [44]:
metric_name = f'{label}'
temp.columns = ['Country', 'Date', metric_name]
# temp.loc[:, 'month'] = temp.date.dt.strftime('%Y-%m')
temp.loc[:, "Total_Infected"] = temp.loc[:, "Total_Infected"].round(0)  
temp.groupby('Country').last()

Date  Total_Infected
Country                                        
Afghanistan          2020-03-28           383.0
Algeria              2020-03-28          1750.0
Argentina            2020-03-28          3018.0
Australia            2020-03-28         13834.0
Austria              2020-03-28         37228.0
...                         ...             ...
United Arab Emirates 2020-03-28          1440.0
United Kingdom       2020-03-28         83653.0
Uzbekistan           2020-03-28           358.0
Venezuela            2020-03-28           267.0
Zimbabwe             2020-03-28            24.0

[80 rows x 2 columns]

## Infected vs. number of confirmed cases
> Allows you to compare how countries have been tracking the true number of infected people. 
The smaller deviation from the dashed line (45 degree line) the better job at tracking the true number of infected people.

In [45]:
data_pc = df_final.copy()

In [46]:
data_countries = []
data_countries_pc = []

In [47]:
for i in countries_n_plus:
    data_pc.loc[data_pc.location == i,"total_infected"] = data_pc.loc[data_pc.location == i,"total_infected"]
    data_pc.loc[data_pc.location == i,"total_cases"] = data_pc.loc[data_pc.location == i,"total_cases"]
    # get each country time series
filter1 = data_pc["total_infected"] > 1
for i in countries_n_plus:
    filter_country = data_pc["location"]== i
    data_countries_pc.append(data_pc[filter_country & filter1])

In [48]:
type(data_countries_pc[0])

pandas.core.frame.DataFrame

In [49]:
data_countries_pc[0]

location       date  total_deaths  total_cases       CFR  \
52  Afghanistan 2020-03-14             0           11  0.000000   
53  Afghanistan 2020-03-15             0           16  0.000000   
54  Afghanistan 2020-03-16             0           21  0.000000   
55  Afghanistan 2020-03-17             0           22  0.000000   
56  Afghanistan 2020-03-18             0           22  0.000000   
57  Afghanistan 2020-03-19             0           22  0.000000   
58  Afghanistan 2020-03-20             0           24  0.000000   
59  Afghanistan 2020-03-21             0           24  0.000000   
60  Afghanistan 2020-03-22             1           40  0.025000   
61  Afghanistan 2020-03-23             1           40  0.025000   
62  Afghanistan 2020-03-24             1           74  0.013514   
63  Afghanistan 2020-03-25             2           84  0.023810   
64  Afghanistan 2020-03-26             4           94  0.042553   
65  Afghanistan 2020-03-27             4          110  0.036364   
66  Afghanistan 2020-03-28             4          110  0.036364   

    total_infected  infected_g  
52       40.000000         NaN  
53       40.000000    0.000000  
54       74.000000    0.615186  
55       84.000000    0.126752  
56       94.000000    0.112478  
57      110.000000    0.157186  
58      110.000000    0.000000  
59      128.546017         NaN  
60      150.218895         NaN  
61      175.545823         NaN  
62      205.142876         NaN  
63      239.729996         NaN  
64      280.148509         NaN  
65      327.381590         NaN  
66      382.578176         NaN

In [50]:
def get_df_country(country):
    for i, df in enumerate(data_countries_pc):
        if len(df.loc[df['location'] == country]):
            print(f'country: {country}, index: {i}')
        

In [51]:
get_df_country('Italy')

country: Italy, index: 47


In [52]:
data_countries_pc[47]

location       date  total_deaths  total_cases       CFR  total_infected  \
5516    Italy 2020-02-13             0            3  0.000000       20.000000   
5517    Italy 2020-02-14             0            3  0.000000       62.000000   
5518    Italy 2020-02-15             0            3  0.000000      155.000000   
5519    Italy 2020-02-16             0            3  0.000000      229.000000   
5520    Italy 2020-02-17             0            3  0.000000      322.000000   
5521    Italy 2020-02-18             0            3  0.000000      453.000000   
5522    Italy 2020-02-19             0            3  0.000000      655.000000   
5523    Italy 2020-02-20             0            3  0.000000      888.000000   
5524    Italy 2020-02-21             1           20  0.050000     1128.000000   
5525    Italy 2020-02-22             2           62  0.032258     1694.000000   
5526    Italy 2020-02-23             3          155  0.019355     2036.000000   
5527    Italy 2020-02-24             7          229  0.030568     2502.000000   
5528    Italy 2020-02-25            10          322  0.031056     3089.000000   
5529    Italy 2020-02-26            12          453  0.026490     3858.000000   
5530    Italy 2020-02-27            17          655  0.025954     4636.000000   
5531    Italy 2020-02-28            21          888  0.023649     5883.000000   
5532    Italy 2020-02-29            29         1128  0.025709     7375.000000   
5533    Italy 2020-03-01            34         1694  0.020071     9172.000000   
5534    Italy 2020-03-02            52         2036  0.025540    10149.000000   
5535    Italy 2020-03-03            79         2502  0.031575    12462.000000   
5536    Italy 2020-03-04           107         3089  0.034639    12462.000000   
5537    Italy 2020-03-05           148         3858  0.038362    17660.000000   
5538    Italy 2020-03-06           197         4636  0.042494    21157.000000   
5539    Italy 2020-03-07           233         5883  0.039606    24747.000000   
5540    Italy 2020-03-08           366         7375  0.049627    27980.000000   
5541    Italy 2020-03-09           463         9172  0.050480    31506.000000   
5542    Italy 2020-03-10           631        10149  0.062174    35713.000000   
5543    Italy 2020-03-11           827        12462  0.066362    41035.000000   
5544    Italy 2020-03-12           827        12462  0.066362    47021.000000   
5545    Italy 2020-03-13          1266        17660  0.071687    53578.000000   
5546    Italy 2020-03-14          1441        21157  0.068110    59138.000000   
5547    Italy 2020-03-15          1809        24747  0.073100    63927.000000   
5548    Italy 2020-03-16          2158        27980  0.077127    69176.000000   
5549    Italy 2020-03-17          2503        31506  0.079445    74386.000000   
5550    Italy 2020-03-18          2978        35713  0.083387    80589.000000   
5551    Italy 2020-03-19          3405        41035  0.082978    86498.000000   
5552    Italy 2020-03-20          4032        47021  0.085749    92472.000000   
5553    Italy 2020-03-21          4825        53578  0.090056   114148.795371   
5554    Italy 2020-03-22          5476        59138  0.092597   140906.950047   
5555    Italy 2020-03-23          6077        63927  0.095062   173937.609302   
5556    Italy 2020-03-24          6820        69176  0.098589   214711.140362   
5557    Italy 2020-03-25          7503        74386  0.100866   265042.586136   
5558    Italy 2020-03-26          8215        80589  0.101937   327172.462254   
5559    Italy 2020-03-27          9134        86498  0.105598   403866.494128   
5560    Italy 2020-03-28         10023        92472  0.108390   498538.733840   

      infected_g  
5516         NaN  
5517    1.131402  
5518    0.916291  
5519    0.390297  
5520    0.340830  
5521    0.341341  
5522    0.368743  
5523    0.304337  
5524    0.239230  
5525    0.406646  
5526    0.183895  
5527    0.206103  
5528    0.210757  
5529    0.22

In [79]:
df_all_data_countries_pc=pd.concat(data_countries_pc)

In [81]:
df_all_data_countries_pc.tail()

location       date  total_deaths  total_cases       CFR  \
11854  Zimbabwe 2020-03-24             1            3  0.333333   
11855  Zimbabwe 2020-03-25             1            3  0.333333   
11856  Zimbabwe 2020-03-26             1            3  0.333333   
11857  Zimbabwe 2020-03-27             1            5  0.200000   
11858  Zimbabwe 2020-03-28             1            7  0.142857   

       total_infected  infected_g  
11854       13.092992         NaN  
11855       15.311724         NaN  
11856       17.906443         NaN  
11857       20.940861         NaN  
11858       24.489490         NaN

In [ ]:
#### save all pred as one df

In [82]:
df_all_data_countries_pc.to_csv(DATA_PARENT/'data/processed/csse/df_all_data_countries_pc.csv')

In [ ]:
### Combine last day only pred with un and freedom house data

In [53]:
df_country_un_stats = pd.read_csv(DATA_PARENT/'data/un/df_un_merged_stats.csv')

In [60]:
df_country_un_stats.rename(columns={'Country': 'location'}, inplace=True)

In [61]:
idx = data_countries_pc[0].groupby(['location'])['date'].transform(max) == data_countries_pc[0]['date']
sub_df=data_countries_pc[0][idx]
sub_df

location       date  total_deaths  total_cases       CFR  \
66  Afghanistan 2020-03-28             4          110  0.036364   

    total_infected  infected_g  
66      382.578176         NaN

In [62]:
sub_df.iloc[0]['location']

'Afghanistan'

In [63]:
df_country_un_stats.head()

location  Population_million  Population_density  Population_60+  \
0  Afghanistan             38.0418             58.2694          4.1655   
1      Albania              2.8809            105.1430         20.4779   
2      Algeria             43.0531             18.0763          9.6817   
3      Andorra              0.0771            164.1319         19.0387   
4       Angola             31.8253             25.5276          3.6243   

   Physicians_per_1000_pop  
0                      0.3  
1                      1.2  
2                      1.8  
3                      3.3  
4                      0.2

In [ ]:
### freedom house

In [72]:
df_freedomhouse_merged = pd.read_csv(DATA_PARENT/'data/freedom_house/df_freedomhouse_merged.csv')

In [73]:
df_freedomhouse_merged.head()

Country  Global_Score Global_Status  Political Rights  Civil Liberties  \
0     Canada            98          Free                40               58   
1  Australia            97          Free                40               57   
2      Japan            96          Free                40               56   
3    Estonia            94          Free                38               56   
4    Germany            94          Free                39               55   

   Internet_Score  Internet_Access_Obstacles  Internet_Content_Limit  \
0              87                         23                      33   
1              77                         23                      29   
2              73                         21                      28   
3              94                         25                      32   
4              80                         22                      30   

   Internet_User_Rights_Violations Internet_Status  
0                               31            Free  
1                               25            Free  
2                               24            Free  
3                               37            Free  
4                               28            Free

In [74]:
df_freedomhouse_merged.rename(columns={'Country': 'location'}, inplace=True)

In [76]:
frames=[]
for df in data_countries_pc:
    idx = df.groupby(['location'])['date'].transform(max) == df['date']
    sub_df=df[idx]
    if len(sub_df)>0:
        #print(f'sub_df: {sub_df}')
        country=sub_df.iloc[0]['location']
        un_df=df_country_un_stats.loc[df_country_un_stats['location'] == country]
        #print(f'un_df: {un_df}')
        df_merged=pd.merge(sub_df, un_df)
        #freedom house data
        fh_df=df_freedomhouse_merged.loc[df_freedomhouse_merged['location'] == country]
        df_merged=pd.merge(df_merged, fh_df)
        frames.append(df_merged)
df_all_un_fh=pd.concat(frames)


In [77]:
df_all_un_fh.head()

location       date  total_deaths  total_cases       CFR  total_infected  \
0   Argentina 2020-03-28            18          690  0.026087     3018.475269   
0   Australia 2020-03-28            14         3640  0.003846    13833.806652   
0  Azerbaijan 2020-03-28             4          182  0.021978      623.509155   
0  Bangladesh 2020-03-28             5           48  0.104167      121.602482   
0      Brazil 2020-03-28           111         3904  0.028432    20072.663760   

   infected_g  Population_million  Population_density  Population_60+  \
0         NaN             44.7807             16.3631         15.3941   
0         NaN             25.2032              3.2807         21.4395   
0         NaN             10.0477            121.5577         11.0734   
0         NaN            163.0462           1252.5633          7.7418   
0         NaN            211.0495             25.2508         13.5897   

   Physicians_per_1000_pop  Global_Score Global_Status  Political Rights  \
0                      4.0            85          Free                35   
0                      3.6            97          Free                40   
0                      3.4            10      Not Free                 2   
0                      0.5            39   Partly Free                15   
0                      2.1            75          Free                31   

   Civil Liberties  Internet_Score  Internet_Access_Obstacles  \
0               50              72                         19   
0               57              77                         23   
0                8              39                         11   
0               24              44                         13   
0               44              64                         18   

   Internet_Content_Limit  Internet_User_Rights_Violations Internet_Status  
0                      26                               27            Free  
0                      29                               25            Free  
0                      15                               13        Not Free  
0                      17                               14     Partly Free  
0                      26                               20     Partly Free

In [78]:
df_all_un_fh.to_csv(DATA_PARENT/'data/processed/csse/df_data_countries_pc_latest.csv')

## Methodology

We argue that the number of infected in the past can be infered using today's number of deaths and average fatality rate from confirmed cases in the following way:

{% raw %}
$$ I_{t-j} = \frac{D_t}{{CFR}_t}$$
{% endraw %}

where {% raw %}$I_t${% endraw %} = number of infected, {% raw %}$D_t${% endraw %} = number of deaths, and {% raw %}${CFR}_t ${% endraw %} = case fatality rate = {% raw %}$\frac{D}{C}${% endraw %}. The {% raw %}$j${% endraw %} depends on the average number of days that covid patients die after having the first symptoms.

**Assumption 1**: The case fatality rate is a good proxy for the fatality rate of the infected population


Then, in order to estimate the current number of infected {% raw %}$I_t${% endraw %} we need to estimate its growth rate from {% raw %}$t-j${% endraw %} to {% raw %}$t${% endraw %}.

{% raw %}
$$I_t = (1+\hat{g})^j I_{t-j}$$
{% endraw %}

**Assumption 2**: The growth rate of infected $\hat{g}$ is an unbiased estimate of $g$ .

For now we estimate $g$ using the average growth rate since having the first infected person.

**Assumption 3**: It takes on average 8 days to day after having the first symptoms.

This analysis was conducted by [Joao B. Duarte](https://www.jbduarte.com). Relevant sources are listed below: 


1. [2019 Novel Coronavirus COVID-19 (2019-nCoV) Data Repository by Johns Hopkins CSSE](https://systems.jhu.edu/research/public-health/ncov/) [GitHub repository](https://github.com/CSSEGISandData/COVID-19). 

2. [Feenstra, Robert C., Robert Inklaar and Marcel P. Timmer (2015), "The Next Generation of the Penn World Table" American Economic Review, 105(10), 3150-3182](https://www.rug.nl/ggdc/productivity/pwt/related-research)
